File: examples/flutter-host/lib/src/player/example_dolby_acceptance_catalog.dart
import 'package:vesper_player/vesper_player.dart';

import '../hdr_evidence/hdr_evidence_capture.dart';

const String exampleDolbyAcceptanceWidevineLicenseUri =
    'https://widevine-dash.ezdrm.com/proxy?pX=E8A6EE';

const List<int> exampleDolbyAcceptanceFpsValues = <int>[24, 30, 50, 120];

enum ExampleDolbyAcceptanceProfile { p5, p81, p84 }

extension ExampleDolbyAcceptanceProfileLabels on ExampleDolbyAcceptanceProfile {
  String get pathSegment {
    return switch (this) {
      ExampleDolbyAcceptanceProfile.p5 => 'p5',
      ExampleDolbyAcceptanceProfile.p81 => 'p81',
      ExampleDolbyAcceptanceProfile.p84 => 'p84',
    };
  }

  String get title {
    return switch (this) {
      ExampleDolbyAcceptanceProfile.p5 => 'P5',
      ExampleDolbyAcceptanceProfile.p81 => 'P8.1',
      ExampleDolbyAcceptanceProfile.p84 => 'P8.4',
    };
  }

  String get sampleIdSegment {
    return switch (this) {
      ExampleDolbyAcceptanceProfile.p5 => 'P5',
      ExampleDolbyAcceptanceProfile.p81 => 'P81',
      ExampleDolbyAcceptanceProfile.p84 => 'P84',
    };
  }

  int get dolbyVisionProfile {
    return switch (this) {
      ExampleDolbyAcceptanceProfile.p5 => 5,
      ExampleDolbyAcceptanceProfile.p81 ||
      ExampleDolbyAcceptanceProfile.p84 => 8,
    };
  }

  String get profileFamily {
    return switch (this) {
      ExampleDolbyAcceptanceProfile.p5 => 'profile5',
      ExampleDolbyAcceptanceProfile.p81 => 'profile8.1',
      ExampleDolbyAcceptanceProfile.p84 => 'profile8.4',
    };
  }

  String get fallbackTarget {
    return switch (this) {
      ExampleDolbyAcceptanceProfile.p5 => 'none',
      ExampleDolbyAcceptanceProfile.p81 => 'hdr10',
      ExampleDolbyAcceptanceProfile.p84 => 'hlg',
    };
  }

  String get transferFunction {
    return switch (this) {
      ExampleDolbyAcceptanceProfile.p84 => 'ARIB_STD_B67_HLG',
      ExampleDolbyAcceptanceProfile.p5 ||
      ExampleDolbyAcceptanceProfile.p81 => 'SMPTE_ST_2084_PQ',
    };
  }
}

enum ExampleDolbyAcceptanceDrmKind { clear, widevine, fairPlayPending }

extension ExampleDolbyAcceptanceDrmKindLabels on ExampleDolbyAcceptanceDrmKind {
  String get title {
    return switch (this) {
      ExampleDolbyAcceptanceDrmKind.clear => 'Clear',
      ExampleDolbyAcceptanceDrmKind.widevine => 'Widevine',
      ExampleDolbyAcceptanceDrmKind.fairPlayPending => 'FairPlay pending',
    };
  }

  String get sampleIdSegment {
    return switch (this) {
      ExampleDolbyAcceptanceDrmKind.clear => 'CLEAR',
      ExampleDolbyAcceptanceDrmKind.widevine => 'WIDEVINE',
      ExampleDolbyAcceptanceDrmKind.fairPlayPending => 'FAIRPLAY-PENDING',
    };
  }

  String get metadataValue {
    return switch (this) {
      ExampleDolbyAcceptanceDrmKind.clear => 'none',
      ExampleDolbyAcceptanceDrmKind.widevine => 'widevine',
      ExampleDolbyAcceptanceDrmKind.fairPlayPending => 'fairPlayPending',
    };
  }
}

final class ExampleDolbyAcceptancePreset {
  const ExampleDolbyAcceptancePreset({
    required this.id,
    required this.label,
    required this.profile,
    required this.fps,
    required this.protocol,
    required this.drmKind,
    required this.source,
    required this.expectedHdrKind,
    required this.manualGate,
    this.notes = const <String>[],
    this.enabled = true,
  });

  final String id;
  final String label;
  final ExampleDolbyAcceptanceProfile profile;
  final int fps;
  final VesperPlayerSourceProtocol protocol;
  final ExampleDolbyAcceptanceDrmKind drmKind;
  final VesperPlayerSource source;
  final String expectedHdrKind;
  final String manualGate;
  final List<String> notes;
  final bool enabled;

  bool get isDrm => drmKind != ExampleDolbyAcceptanceDrmKind.clear;

  bool get isPlayable =>
      enabled && drmKind != ExampleDolbyAcceptanceDrmKind.fairPlayPending;

  String get protocolLabel {
    return switch (protocol) {
      VesperPlayerSourceProtocol.dash => 'DASH',
      VesperPlayerSourceProtocol.hls => 'HLS',
      _ => protocol.name,
    };
  }

  ExampleHdrEvidenceSamplePreset toHdrEvidencePreset() {
    return ExampleHdrEvidenceSamplePreset(
      sampleId: id,
      label: label,
      expectedAxis: 'display',
      sourceMetadata: <String, Object?>{
        'sourceUri': source.uri,
        'sourceKind': 'remote',
        'container': protocolLabel.toLowerCase(),
        'manifestKind': protocolLabel.toLowerCase(),
        'codec': 'dolby-vision',
        'sampleMimeType': 'video/dolby-vision',
        'width': null,
        'height': null,
        'frameRate': fps.toDouble(),
        'bitDepth': 10,
        'hdrKind': expectedHdrKind,
        'colorPrimaries': 'BT.2020',
        'transferFunction': profile.transferFunction,
        'yCbCrMatrix': 'BT.2020_NCL',
        'drmKind': drmKind.metadataValue,
        'manualGate': manualGate,
        'controlPurpose': 'dolbyVisionAcceptance',
        'dolbyVision': <String, Object?>{
          'profile': profile.dolbyVisionProfile,
          'profileFamily': profile.profileFamily,
          'baseLayer': 'hevc-main10',
          'fallbackTarget': profile.fallbackTarget,
          'containerEvidence': 'dolby-browser-test-kit',
        },
        'metadataTool': <String, Object?>{
          'name': 'Dolby Browser Test Kit',
          'version': 'public',
          'command': 'catalog-url',
        },
        'notes': <String>[
          'Dolby Browser Test Kit public URL; media is not bundled.',
          ...notes,
        ],
      },
    );
  }
}

String exampleDolbyAcceptanceUrl({
  required ExampleDolbyAcceptanceProfile profile,
  required int fps,
  required VesperPlayerSourceProtocol protocol,
  required ExampleDolbyAcceptanceDrmKind drmKind,
}) {
  final protocolFile = switch (protocol) {
    VesperPlayerSourceProtocol.dash => 'dash.mpd',
    VesperPlayerSourceProtocol.hls => 'master.m3u8',
    _ => throw ArgumentError.value(protocol, 'protocol', 'DASH or HLS only'),
  };
  final pathKind = switch (drmKind) {
    ExampleDolbyAcceptanceDrmKind.clear => 'clear',
    ExampleDolbyAcceptanceDrmKind.widevine => 'cenc',
    ExampleDolbyAcceptanceDrmKind.fairPlayPending => 'cbcs',
  };
  return 'https://ott.dolby.com/browser_test_kit/$pathKind/'
      '${profile.pathSegment}/$fps/$protocolFile';
}

List<ExampleDolbyAcceptancePreset> buildExampleDolbyAcceptanceCatalog() {
  return <ExampleDolbyAcceptancePreset>[
    for (final profile in ExampleDolbyAcceptanceProfile.values)
      for (final fps
          in exampleDolbyAcceptanceFpsValues) ...<ExampleDolbyAcceptancePreset>[
        _buildDolbyPreset(
          profile: profile,
          fps: fps,
          protocol: VesperPlayerSourceProtocol.dash,
          drmKind: ExampleDolbyAcceptanceDrmKind.clear,
        ),
        _buildDolbyPreset(
          profile: profile,
          fps: fps,
          protocol: VesperPlayerSourceProtocol.hls,
          drmKind: ExampleDolbyAcceptanceDrmKind.clear,
        ),
        _buildDolbyPreset(
          profile: profile,
          fps: fps,
          protocol: VesperPlayerSourceProtocol.dash,
          drmKind: ExampleDolbyAcceptanceDrmKind.widevine,
        ),
        _buildDolbyPreset(
          profile: profile,
          fps: fps,
          protocol: VesperPlayerSourceProtocol.hls,
          drmKind: ExampleDolbyAcceptanceDrmKind.fairPlayPending,
          enabled: false,
        ),
      ],
  ];
}

final List<ExampleDolbyAcceptancePreset> exampleDolbyAcceptanceCatalog =
    buildExampleDolbyAcceptanceCatalog();

ExampleDolbyAcceptancePreset? exampleDolbyAcceptancePresetById(String id) {
  for (final preset in exampleDolbyAcceptanceCatalog) {
    if (preset.id == id) {
      return preset;
    }
  }
  return null;
}

List<ExampleHdrEvidenceSamplePreset>
exampleDolbyAcceptanceHdrEvidencePresets() {
  return exampleDolbyAcceptanceCatalog
      .where((preset) => preset.isPlayable)
      .map((preset) => preset.toHdrEvidencePreset())
      .toList(growable: false);
}

ExampleDolbyAcceptancePreset _buildDolbyPreset({
  required ExampleDolbyAcceptanceProfile profile,
  required int fps,
  required VesperPlayerSourceProtocol protocol,
  required ExampleDolbyAcceptanceDrmKind drmKind,
  bool enabled = true,
}) {
  final protocolSegment = switch (protocol) {
    VesperPlayerSourceProtocol.dash => 'DASH',
    VesperPlayerSourceProtocol.hls => 'HLS',
    _ => throw ArgumentError.value(protocol, 'protocol', 'DASH or HLS only'),
  };
  final id =
      'DOLBY-DV-${profile.sampleIdSegment}-$fps-'
      '$protocolSegment-${drmKind.sampleIdSegment}';
  final label = '${profile.title} ${fps}fps $protocolSegment ${drmKind.title}';
  final uri = exampleDolbyAcceptanceUrl(
    profile: profile,
    fps: fps,
    protocol: protocol,
    drmKind: drmKind,
  );
  final source = VesperPlayerSource(
    uri: uri,
    label: label,
    kind: VesperPlayerSourceKind.remote,
    protocol: protocol,
    drmConfiguration: drmKind == ExampleDolbyAcceptanceDrmKind.widevine
        ? const VesperPlayerDrmConfiguration(
            keySystem: 'widevine',
            licenseUri: exampleDolbyAcceptanceWidevineLicenseUri,
          )
        : null,
  );
  final notes = <String>[
    if (drmKind == ExampleDolbyAcceptanceDrmKind.widevine)
      'Widevine DASH direct native route only.',
    if (drmKind == ExampleDolbyAcceptanceDrmKind.fairPlayPending)
      'FairPlay certificate URI/base64 is not available yet; preset is disabled.',
    if (fps == 50) 'Dolby 50fps signal covers the 60-ish validation bucket.',
    'MP4 zip assets remain manual local-file material and are not bundled.',
  ];
  return ExampleDolbyAcceptancePreset(
    id: id,
    label: label,
    profile: profile,
    fps: fps,
    protocol: protocol,
    drmKind: drmKind,
    source: source,
    expectedHdrKind: 'dolbyVision',
    manualGate: 'requiresDolbyVisionDisplay',
    notes: notes,
    enabled: enabled,
  );
}

File: examples/android-compose-host/app/src/main/java/io/github/umbrella22/vesper/example/androidcomposehost/ExampleDolbyAcceptanceCatalog.kt
package io.github.umbrella22.vesper.example.androidcomposehost

import io.github.umbrella22.vesper.player.android.VesperPlayerDrmConfiguration
import io.github.umbrella22.vesper.player.android.VesperPlayerSource
import io.github.umbrella22.vesper.player.android.VesperPlayerSourceKind
import io.github.umbrella22.vesper.player.android.VesperPlayerSourceProtocol


internal const val EXAMPLE_DOLBY_ACCEPTANCE_WIDEVINE_LICENSE_URI: String =
    "https://widevine-dash.ezdrm.com/proxy?pX=E8A6EE"

internal val exampleDolbyAcceptanceFpsValues: List<Int> = listOf(24, 30, 50, 120)

internal enum class ExampleDolbyAcceptanceProfile(
    val pathSegment: String,
    val title: String,
    val sampleIdSegment: String,
    val dolbyVisionProfile: Int,
    val profileFamily: String,
    val fallbackTarget: String,
    val transferFunction: String,
) {
    P5(
        pathSegment = "p5",
        title = "P5",
        sampleIdSegment = "P5",
        dolbyVisionProfile = 5,
        profileFamily = "profile5",
        fallbackTarget = "none",
        transferFunction = "SMPTE_ST_2084_PQ",
    ),
    P81(
        pathSegment = "p81",
        title = "P8.1",
        sampleIdSegment = "P81",
        dolbyVisionProfile = 8,
        profileFamily = "profile8.1",
        fallbackTarget = "hdr10",
        transferFunction = "SMPTE_ST_2084_PQ",
    ),
    P84(
        pathSegment = "p84",
        title = "P8.4",
        sampleIdSegment = "P84",
        dolbyVisionProfile = 8,
        profileFamily = "profile8.4",
        fallbackTarget = "hlg",
        transferFunction = "ARIB_STD_B67_HLG",
    ),
}

internal enum class ExampleDolbyAcceptanceDrmKind(
    val title: String,
    val sampleIdSegment: String,
    val metadataValue: String,
) {
    Clear(
        title = "Clear",
        sampleIdSegment = "CLEAR",
        metadataValue = "none",
    ),
    Widevine(
        title = "Widevine",
        sampleIdSegment = "WIDEVINE",
        metadataValue = "widevine",
    ),
    FairPlayPending(
        title = "FairPlay pending",
        sampleIdSegment = "FAIRPLAY-PENDING",
        metadataValue = "fairPlayPending",
    ),
}

internal data class ExampleDolbyAcceptancePreset(
    val id: String,
    val label: String,
    val profile: ExampleDolbyAcceptanceProfile,
    val fps: Int,
    val protocol: VesperPlayerSourceProtocol,
    val drmKind: ExampleDolbyAcceptanceDrmKind,
    val source: VesperPlayerSource,
    val expectedHdrKind: String,
    val manualGate: String,
    val notes: List<String> = emptyList(),
    val enabled: Boolean = true,
) {
    val isDrm: Boolean
        get() = drmKind != ExampleDolbyAcceptanceDrmKind.Clear

    val isPlayable: Boolean
        get() = enabled && drmKind != ExampleDolbyAcceptanceDrmKind.FairPlayPending

    val protocolLabel: String
        get() =
            when (protocol) {
                VesperPlayerSourceProtocol.Dash -> "DASH"
                VesperPlayerSourceProtocol.Hls -> "HLS"
                else -> protocol.name
            }

    fun toHdrEvidencePreset(): ExampleHdrEvidenceSamplePreset =
        ExampleHdrEvidenceSamplePreset(
            sampleId = id,
            label = label,
            expectedAxis = "display",
            sourceMetadata =
                mapOf(
                    "sourceUri" to source.uri,
                    "sourceKind" to "remote",
                    "container" to protocolLabel.lowercase(),
                    "manifestKind" to protocolLabel.lowercase(),
                    "codec" to "dolby-vision",
                    "sampleMimeType" to "video/dolby-vision",
                    "width" to null,
                    "height" to null,
                    "frameRate" to fps.toDouble(),
                    "bitDepth" to 10,
                    "hdrKind" to expectedHdrKind,
                    "colorPrimaries" to "BT.2020",
                    "transferFunction" to profile.transferFunction,
                    "yCbCrMatrix" to "BT.2020_NCL",
                    "drmKind" to drmKind.metadataValue,
                    "manualGate" to manualGate,
                    "controlPurpose" to "dolbyVisionAcceptance",
                    "dolbyVision" to
                        mapOf(
                            "profile" to profile.dolbyVisionProfile,
                            "profileFamily" to profile.profileFamily,
                            "baseLayer" to "hevc-main10",
                            "fallbackTarget" to profile.fallbackTarget,
                            "containerEvidence" to "dolby-browser-test-kit",
                        ),
                    "metadataTool" to
                        mapOf(
                            "name" to "Dolby Browser Test Kit",
                            "version" to "public",
                            "command" to "catalog-url",
                        ),
                    "notes" to
                        listOf(
                            "Dolby Browser Test Kit public URL; media is not bundled.",
                        ) + notes,
                ),
        )
}

internal fun exampleDolbyAcceptanceUrl(
    profile: ExampleDolbyAcceptanceProfile,
    fps: Int,
    protocol: VesperPlayerSourceProtocol,
    drmKind: ExampleDolbyAcceptanceDrmKind,
): String {
    val protocolFile =
        when (protocol) {
            VesperPlayerSourceProtocol.Dash -> "dash.mpd"
            VesperPlayerSourceProtocol.Hls -> "master.m3u8"
            else -> error("DASH or HLS only")
        }
    val pathKind =
        when (drmKind) {
            ExampleDolbyAcceptanceDrmKind.Clear -> "clear"
            ExampleDolbyAcceptanceDrmKind.Widevine -> "cenc"
            ExampleDolbyAcceptanceDrmKind.FairPlayPending -> "cbcs"
        }
    return "https://ott.dolby.com/browser_test_kit/$pathKind/" +
        "${profile.pathSegment}/$fps/$protocolFile"
}

internal fun buildExampleDolbyAcceptanceCatalog(): List<ExampleDolbyAcceptancePreset> =
    buildList {
        ExampleDolbyAcceptanceProfile.values().forEach { profile ->
            exampleDolbyAcceptanceFpsValues.forEach { fps ->
                add(
                    buildExampleDolbyAcceptancePreset(
                        profile = profile,
                        fps = fps,
                        protocol = VesperPlayerSourceProtocol.Dash,
                        drmKind = ExampleDolbyAcceptanceDrmKind.Clear,
                    ),
                )
                add(
                    buildExampleDolbyAcceptancePreset(
                        profile = profile,
                        fps = fps,
                        protocol = VesperPlayerSourceProtocol.Hls,
                        drmKind = ExampleDolbyAcceptanceDrmKind.Clear,
                    ),
                )
                add(
                    buildExampleDolbyAcceptancePreset(
                        profile = profile,
                        fps = fps,
                        protocol = VesperPlayerSourceProtocol.Dash,
                        drmKind = ExampleDolbyAcceptanceDrmKind.Widevine,
                    ),
                )
                add(
                    buildExampleDolbyAcceptancePreset(
                        profile = profile,
                        fps = fps,
                        protocol = VesperPlayerSourceProtocol.Hls,
                        drmKind = ExampleDolbyAcceptanceDrmKind.FairPlayPending,
                        enabled = false,
                    ),
                )
            }
        }
    }

internal val exampleDolbyAcceptanceCatalog: List<ExampleDolbyAcceptancePreset> =
    buildExampleDolbyAcceptanceCatalog()

internal fun exampleDolbyAcceptancePresetById(id: String): ExampleDolbyAcceptancePreset? =
    exampleDolbyAcceptanceCatalog.firstOrNull { preset -> preset.id == id }

internal fun filterDolbyAcceptancePresets(
    presets: List<ExampleDolbyAcceptancePreset>,
    drmKind: ExampleDolbyAcceptanceDrmKind,
    profile: ExampleDolbyAcceptanceProfile?,
    fps: Int?,
): List<ExampleDolbyAcceptancePreset> =
    presets.filter { preset ->
        preset.drmKind == drmKind &&
            (profile == null || preset.profile == profile) &&
            (fps == null || preset.fps == fps)
    }

internal fun exampleDolbyAcceptanceHdrEvidencePresets(): List<ExampleHdrEvidenceSamplePreset> =
    exampleDolbyAcceptanceCatalog
        .filter { preset -> preset.isPlayable }
        .map { preset -> preset.toHdrEvidencePreset() }

private fun buildExampleDolbyAcceptancePreset(
    profile: ExampleDolbyAcceptanceProfile,
    fps: Int,
    protocol: VesperPlayerSourceProtocol,
    drmKind: ExampleDolbyAcceptanceDrmKind,
    enabled: Boolean = true,
): ExampleDolbyAcceptancePreset {
    val protocolSegment =
        when (protocol) {
            VesperPlayerSourceProtocol.Dash -> "DASH"
            VesperPlayerSourceProtocol.Hls -> "HLS"
            else -> error("DASH or HLS only")
        }
    val id =
        "DOLBY-DV-${profile.sampleIdSegment}-$fps-" +
            "$protocolSegment-${drmKind.sampleIdSegment}"
    val label = "${profile.title} ${fps}fps $protocolSegment ${drmKind.title}"
    val uri =
        exampleDolbyAcceptanceUrl(
            profile = profile,
            fps = fps,
            protocol = protocol,
            drmKind = drmKind,
        )
    val source =
        VesperPlayerSource(
            uri = uri,
            label = label,
            kind = VesperPlayerSourceKind.Remote,
            protocol = protocol,
            drmConfiguration =
                if (drmKind == ExampleDolbyAcceptanceDrmKind.Widevine) {
                    VesperPlayerDrmConfiguration(
                        keySystem = "widevine",
                        licenseUri = EXAMPLE_DOLBY_ACCEPTANCE_WIDEVINE_LICENSE_URI,
                    )
                } else {
                    null
                },
        )
    val notes =
        buildList {
            if (drmKind == ExampleDolbyAcceptanceDrmKind.Widevine) {
                add("Widevine DASH direct native route only.")
            }
            if (drmKind == ExampleDolbyAcceptanceDrmKind.FairPlayPending) {
                add("FairPlay certificate URI/base64 is not available yet; preset is disabled.")
            }
            if (fps == 50) {
                add("Dolby 50fps signal covers the 60-ish validation bucket.")
            }
            add("MP4 zip assets remain manual local-file material and are not bundled.")
        }
    return ExampleDolbyAcceptancePreset(
        id = id,
        label = label,
        profile = profile,
        fps = fps,
        protocol = protocol,
        drmKind = drmKind,
        source = source,
        expectedHdrKind = "dolbyVision",
        manualGate = "requiresDolbyVisionDisplay",
        notes = notes,
        enabled = enabled,
    )
}
